### ISSUE TITLE
Rule chaining

### ISSUE THREAD

**Issue by damianon** [mozilla-iot/gateway#2190]
I want to be able to create a rule similar to this:

At 8PM turn off light X and set light Y to color <dark>. And if light X was on, turn on light Y.

I haven't found a way to do this in one rule. Two rules could be possible, but with that approach the rules view would become very full and cluttered very fast and it's not easily visible which rules belong to the same "routine"

**Comment by mrstegeman** [comment]
What you're asking for isn't super clear to me. Can you tell me if I have the idea right?

```
IF (time == 8pm AND lightX.on == TRUE) THEN
    lightX.on = FALSE
    lightY.on = TRUE
    lightY.color = <some color>
END
```

**Comment by damianon** [comment]
Well the exact use case I'd like to cover right now would be 

`IF (time == 8pm) THEN
    IF (lightX.on == TRUE)
        lightY.on = TRUE
        lightY.color = <some color>
    END
    lightX.on = FALSE
END`

The color could be set regardless. I just want to prevent lightY turning on if no light was on in the first place (e.g. we aren't home or in another room)

**Comment by mrstegeman** [comment]
> The color could be set regardless. I just want to prevent lightY turning on if no light was on in the first place (e.g. we aren't home or in another room)

There's no need to set the color if the light won't be turning on. I think my rule above should actually handle your case. lightY will only be turned on if lightX was already on.

**Comment by damianon** [comment]
well. my logic failed me. thanks for that solution

**Comment by damianon** [comment]
Sorry, I just remembered something: 

If lightX is off and lightY is on, the color should still be changed

**Comment by mrstegeman** [comment]
I think what you really want is:

```
IF (time == 8pm) THEN
    IF (lightX.on == TRUE OR lightY.on == TRUE)
        lightX.on = FALSE
        lightY.on = TRUE
        lightY.color = <some color>
    END
END
```

Either way, you're right that you'd currently need multiple rules to handle it, or write an external script using the Web Thing REST API to do this for you.

**Comment by flatsiedatsie** [comment]
@mrstegeman In theory, from version 0.10 onwards, could someone write an add-on that adds a more complex rule system? For example, one in which people could code or use blockly? Not volunteering, just curious :-)

**Comment by mrstegeman** [comment]
Yes, in theory. You'd still have to use the manual OAuth token process, but otherwise, it should work fine.

**Comment by flatsiedatsie** [comment]
`You'd still have to use the manual OAuth token process`

I'm sure version 0.11 will remove that obstacle ;-)

**Comment by damooooooooooh** [comment]
Hi so the Gateway rule engine wont be enhanced to support these types of use cases in the future and rule processing should be triggered outside the gateway?

**Comment by mrstegeman** [comment]
In most cases, these complex rules can be broken down into a set of simpler rules, so the functionality is already available, albeit in an inconvenient fashion.

**Comment by damooooooooooh** [comment]
Got it thanks.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by madb1lly** [comment]
> 
> 
> In most cases, these complex rules can be broken down into a set of simpler rules, so the functionality is already available, albeit in an inconvenient fashion.

Hello @mrstegeman, the inconvenience is the problem, actually, as it makes the rules page very cluttered. Simple rules would be fine if we had a way to visually group or link them, but they are just listed one after the other so it quickly becomes difficult to keep track of which rules are associated with which other ones.

Regarding #2216 and this issue the inability to use more complex logic means that we have no choice but to use simple rules.

So, we need either:
1. Rules can use more complex logic
2. Rules pages allows visual grouping or linking of rules (probably need to raise a new issue for this request).

**Comment by madb1lly** [comment]
@benfrancis considering your comment on #2337 I think my suggested option 2 may be your preferred option:
2. Rules pages allows visual grouping or linking of rules.

**Comment by benfrancis** [comment]
Hi @madb1lly, 

A long-running feature request is [groups for things](https://github.com/mozilla-iot/gateway/issues/1191), so I think we should probably tackle that first.

I'm not so sure about grouping rules, but a first step could be to allow users to re-order rules in the same way that you can already re-order things. That way you can at least put related rules next to each other.

Again, I want to be careful about getting the right balance between simplicity and flexibility for the default rules engine UI. But there's nothing to stop you creating an add-on which provides a more sophisticated view for rules or allows more complex rules to be created.

**Comment by madb1lly** [comment]
> Hi @madb1lly,
> 
> A long-running feature request is [groups for things](https://github.com/mozilla-iot/gateway/issues/1191), so I think we should probably tackle that first.
> 
> I'm not so sure about grouping rules, but a first step could be to allow users to re-order rules in the same way that you can already re-order things. That way you can at least put related rules next to each other.
> 
>

@benfrancis being able to reorder rules will help. Grouping things will be very useful but won't help this issue, will it?
